plugin/pkg/admission/priority/admission.go

```diff
@@ -166,6 +155,11 @@ func (p *PriorityPlugin) admitPod(a admission.Attributes) error {
 	}
 	if utilfeature.DefaultFeatureGate.Enabled(features.PodPriority) {
 		var priority int32
+		// TODO: @ravig - This is for backwards compatibility to ensure that critical pods with annotations just work fine.
+		// Remove when no longer needed.
+		if len(pod.Spec.PriorityClassName) == 0 && kubelettypes.IsCritical(pod.Namespace, pod.Annotations) {
```

Comment:
You can't trust the `pod.Namespace` here.  The user could have lied in the object he is creating.  You need to check the attributes namespace.

Reply:
> You can't trust the pod.Namespace here. The user could have lied in the object he is creating. You need to check the attributes namespace.

Since this is commonly confused: we cannot ensure the well-formedness of the object's namespace before mutation admission because a mutating admission plugin can change it.  We can only confirm and/or force the namespace after mutation is completed.  Another admission plugin could later "make valid" this field (though they shouldn't) and you'd end up having set this for that user.

@kubernetes/sig-api-machinery-bugs an edge we care about?  I suppose it is rather unlikely....

Reply:
Thanks for the awesome explanation @deads2k. I have changed the code to get namepsace using  `a.GetNamespace()`.